<h2 style="text-align: center;" > <strong>SBA Loand Default Prediction</strong></h2>

### What is the SBA?
The SBA was created in 1953 and is a U.S. federal government agency tasked with providing counseling, capital and contracting expertise for entrepreneurs. SBA loans are just a part of what it offers. Small business owners can also get free counseling through resource partners such as Small Business Development Centers, SCORE, Veteran’s Business Centers and Women’s Business Centers. It also provides assistance and expertise for businesses that want to qualify for government contracts or export to other countries. It is funded by taxpayers through Congressional appropriations. That means your tax dollars help it help small business owners, so be sure you take advantage of what it has to offer.

Small businesses have been the primary source of employment in the United States. Helping small businesses help with job creation, which reduces unemployment. Small business growth also promotes economic growth. One of the ways the SBA helps small businesses is by guaranteeing bank loans. This guarantee reduces the risk to banks and encourages them to lend to small businesses. If the loan defaults, the SBA covers the amount guaranteed, and the bank suffers a loss for the remaining balance.

### Dataset
The dataset that I used in this analyst come from [Kaggle!](https://www.kaggle.com/mirbektoktogaraev/should-this-loan-be-approved-or-denied). This dataset contains 27 features with some categorical and numerical features. The target was MIS_Status which showed loan status (CHGOFF or PIF)

### Problem
The default rate for the 10-year period ending in 2008, when nearly 1 in 4 (24.7%) SBA loans weren’t paid back. It means that SBA would have loss due to Small Business couldn’t make repayment

### Goals
Being able to early detected of SBA loan default could reduce the risk of loss for Small Business Administration.



### Step of Work
**Step 1** : In this step, I would to clean the dataset includes handling missing values, subset data, feature engineering and feature selection for next step

In [1]:
# import common pakages
import pandas as pd
import numpy as np
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
import matplotlib.pyplot as plt
import sidetable

%matplotlib inline

In [2]:
df = pd.read_csv(r'SBAnational.csv')
df.head()

LoanNr_ChkDgt                           Name          City State    Zip  \
0     1000014003                 ABC HOBBYCRAFT    EVANSVILLE    IN  47711   
1     1000024006    LANDMARK BAR & GRILLE (THE)     NEW PARIS    IN  46526   
2     1000034009          WHITLOCK DDS, TODD M.   BLOOMINGTON    IN  47401   
3     1000044001  BIG BUCKS PAWN & JEWELRY, LLC  BROKEN ARROW    OK  74012   
4     1000054004    ANASTASIA CONFECTIONS, INC.       ORLANDO    FL  32801   

                            Bank BankState   NAICS ApprovalDate ApprovalFY  \
0               FIFTH THIRD BANK        OH  451120    28-Feb-97       1997   
1                1ST SOURCE BANK        IN  722410    28-Feb-97       1997   
2        GRANT COUNTY STATE BANK        IN  621210    28-Feb-97       1997   
3  1ST NATL BK & TR CO OF BROKEN        OK       0    28-Feb-97       1997   
4        FLORIDA BUS. DEVEL CORP        FL       0    28-Feb-97       1997   

   ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  DisbursementGross  \
0  ...          N       Y         NaN         28-Feb-99        $60,000.00    
1  ...          N       Y         NaN         31-May-97        $40,000.00    
2  ...          N       N         NaN         31-Dec-97       $287,000.00    
3  ...          N       Y         NaN         30-Jun-97        $35,000.00    
4  ...          N       N         NaN         14-May-97       $229,000.00    

   BalanceGross  MIS_Status ChgOffPrinGr        GrAppv      SBA_Appv  
0        $0.00        P I F       $0.00    $60,000.00    $48,000.00   
1        $0.00        P I F       $0.00    $40,000.00    $32,000.00   
2        $0.00        P I F       $0.00   $287,000.00   $215,250.00   
3        $0.00        P I F       $0.00    $35,000.00    $28,000.00   
4        $0.00        P I F       $0.00   $229,000.00   $229,000.00   

[5 rows x 27 columns]

In [3]:
df.dtypes

LoanNr_ChkDgt          int64
Name                  object
City                  object
State                 object
Zip                    int64
Bank                  object
BankState             object
NAICS                  int64
ApprovalDate          object
ApprovalFY            object
Term                   int64
NoEmp                  int64
NewExist             float64
CreateJob              int64
RetainedJob            int64
FranchiseCode          int64
UrbanRural             int64
RevLineCr             object
LowDoc                object
ChgOffDate            object
DisbursementDate      object
DisbursementGross     object
BalanceGross          object
MIS_Status            object
ChgOffPrinGr          object
GrAppv                object
SBA_Appv              object
dtype: object

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 899164 entries, 0 to 899163
Data columns (total 27 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   LoanNr_ChkDgt      899164 non-null  int64  
 1   Name               899150 non-null  object 
 2   City               899134 non-null  object 
 3   State              899150 non-null  object 
 4   Zip                899164 non-null  int64  
 5   Bank               897605 non-null  object 
 6   BankState          897598 non-null  object 
 7   NAICS              899164 non-null  int64  
 8   ApprovalDate       899164 non-null  object 
 9   ApprovalFY         899164 non-null  object 
 10  Term               899164 non-null  int64  
 11  NoEmp              899164 non-null  int64  
 12  NewExist           899028 non-null  float64
 13  CreateJob          899164 non-null  int64  
 14  RetainedJob        899164 non-null  int64  
 15  FranchiseCode      899164 non-null  int64  
 16  Ur

### Missing Values

In [5]:
## Checking missing value
(((df.isna().sum())/len(df))*100).round(4)

LoanNr_ChkDgt         0.0000
Name                  0.0016
City                  0.0033
State                 0.0016
Zip                   0.0000
Bank                  0.1734
BankState             0.1742
NAICS                 0.0000
ApprovalDate          0.0000
ApprovalFY            0.0000
Term                  0.0000
NoEmp                 0.0000
NewExist              0.0151
CreateJob             0.0000
RetainedJob           0.0000
FranchiseCode         0.0000
UrbanRural            0.0000
RevLineCr             0.5036
LowDoc                0.2872
ChgOffDate           81.9055
DisbursementDate      0.2634
DisbursementGross     0.0000
BalanceGross          0.0000
MIS_Status            0.2221
ChgOffPrinGr          0.0000
GrAppv                0.0000
SBA_Appv              0.0000
dtype: float64

### Handling Missing Values, Feature Engineering, Feature Selection

### Name, City, State, Zip

In [6]:
## Handling Missing Value

# Name, City, State and Zip columns is Borrower information
# Name columns has 0.0016 missing values, in this column I would not impute since it was unique and I would not use the column for analyst
# So, I drop Name column
df.drop(columns='Name', inplace=True)

In [7]:
## Handling Missing Values

# For City and State columns has missing values each 0.0033 and 0.0016
# Zip columns has no missing values
# I would imputes City and State columns using Zip
# Since City and State has object type, I would use mode

df['City'] = df.groupby('Zip')['City'].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else 'Empty'))
df['State'] = df.groupby('Zip')['State'].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else 'Empty'))


In [8]:
## lets check for City and State after imputation
df[(df['City'] == 'Empty') | (df['State'] == 'Empty') ]

LoanNr_ChkDgt              City  State    Zip  \
306274     3113583009  BOX 267, APO  AP  Empty  96205   

                             Bank BankState   NAICS ApprovalDate ApprovalFY  \
306274  RTC/WESTPORT SAVINGS BANK        CA  541511    21-Jan-88       1988   

        Term  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
306274    73  ...          N       N   18-Sep-93         16-Mar-88   

        DisbursementGross  BalanceGross MIS_Status ChgOffPrinGr       GrAppv  \
306274        $50,000.00         $0.00      CHGOFF  $30,589.00   $50,000.00    

           SBA_Appv  
306274  $45,000.00   

[1 rows x 26 columns]

In [9]:
df.shape

(899164, 26)

In [10]:
## Since 1 row contain 'Empty', means that there isno information about the State
## I would delete the row

df = df[df['State'] != 'Empty']

df.shape

(899163, 26)

In [ ]:
# sample df1 to impute missing values

# df1={'CIK': ['C','C','B','B','A', 'A', 'C', 'B', 'C','A', 'D', 'D', 'D'],
#     'SIK': ['2','1','4',np.nan,np.nan, '6', '2', '4', '3','6', np.nan, np.nan, np.nan]}

# df1 = pd.DataFrame(data=df1)
# df1['SIK'] = df1.groupby('CIK')['SIK'].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else 'Empty'))
# df1

### NewExist

In [11]:
## Handling Missing Values

# NewExist 
# For actually, we just has binary for NewExist feature (1 = Existing Business, 2 = New Business)
# In fact, there 0 values that should be handle beside NaN
# I assume that we could replace 0 and impute NaN using RetainedJobs
# If RetainedJob > 0 ==> Existing Business (1)

# If NewExist == 0 & RetainedJob>0 ==> 1
# Looping 1
cond = df[(df['NewExist'] == 0) & (df['RetainedJob']>0)].index
for i in cond:
    df.loc[i, ['NewExist']] = 1

# If NewExist == Nan & RetainedJob>0 ==> 1  
# Looping 2
cond1 = df[(df['NewExist'].isna()) & (df['RetainedJob']>0)].index
for i in cond1:
    df.loc[i, ['NewExist']] = 1
    
print('-'*50)
print(df['NewExist'].isna().sum()) #Nan 19 row
print(df['NewExist'].value_counts()) # 0 881 row

--------------------------------------------------
19
1.0    645138
2.0    253125
0.0       881
Name: NewExist, dtype: int64


After some imputation on NewExist feature, we still have 19 NaN. There is no way I could impute with another feature again. So, I decided to drop row of NewExist that contains with NaN

In [12]:
## Handling Missing Values

df.dropna(subset=['NewExist'], inplace=True)

### ChgOffDate

In [13]:
## Handling Missing Value

# Since ChgOffDate has more than 80% Missing value, I would drop this column
df.drop(columns='ChgOffDate', inplace=True)

### MIS_Status

In [14]:
## Handling Missing Values
## Since this feature is the target, I would not impute 

df.dropna(subset=['MIS_Status'], inplace=True)

In [15]:
df['MIS_Status'].isna().sum()

0

Next, let's check for the types of features

In [16]:
# Check data types
df.dtypes

LoanNr_ChkDgt          int64
City                  object
State                 object
Zip                    int64
Bank                  object
BankState             object
NAICS                  int64
ApprovalDate          object
ApprovalFY            object
Term                   int64
NoEmp                  int64
NewExist             float64
CreateJob              int64
RetainedJob            int64
FranchiseCode          int64
UrbanRural             int64
RevLineCr             object
LowDoc                object
DisbursementDate      object
DisbursementGross     object
BalanceGross          object
MIS_Status            object
ChgOffPrinGr          object
GrAppv                object
SBA_Appv              object
dtype: object

I would like to change some features that should be floats but read as objects

In [17]:
df[['DisbursementGross', 'BalanceGross', 'ChgOffPrinGr', 'GrAppv','SBA_Appv']].head()

DisbursementGross BalanceGross ChgOffPrinGr        GrAppv      SBA_Appv
0       $60,000.00        $0.00        $0.00    $60,000.00    $48,000.00 
1       $40,000.00        $0.00        $0.00    $40,000.00    $32,000.00 
2      $287,000.00        $0.00        $0.00   $287,000.00   $215,250.00 
3       $35,000.00        $0.00        $0.00    $35,000.00    $28,000.00 
4      $229,000.00        $0.00        $0.00   $229,000.00   $229,000.00

Let's removing '$' and ',' that caused data types object

In [18]:
## I would replace '$' and ',' from the data and change to float

# DisbursementGross
df['DisbursementGross'] = df['DisbursementGross'].apply(lambda x:x.replace('$', ''))
df['DisbursementGross'] = df['DisbursementGross'].apply(lambda x:x.replace(',', ''))

# BalanceGross
df['BalanceGross'] = df['BalanceGross'].apply(lambda x:x.replace('$', ''))
df['BalanceGross'] = df['BalanceGross'].apply(lambda x:x.replace(',', ''))

# ChgOffPrinGr
df['ChgOffPrinGr'] = df['ChgOffPrinGr'].apply(lambda x:x.replace('$', ''))
df['ChgOffPrinGr'] = df['ChgOffPrinGr'].apply(lambda x:x.replace(',', ''))

# GrAppv
df['GrAppv'] = df['GrAppv'].apply(lambda x:x.replace('$', ''))
df['GrAppv'] = df['GrAppv'].apply(lambda x:x.replace(',', ''))

# SBA_Appv
df['SBA_Appv'] = df['SBA_Appv'].apply(lambda x:x.replace('$', ''))
df['SBA_Appv'] = df['SBA_Appv'].apply(lambda x:x.replace(',', ''))


### ApprovalFY

Now, lets look for ApprovalFY features

In [19]:
# Unique value of ApprovalFY
df['ApprovalFY'].unique()

array([1997, 1980, 2006, 1998, 1999, 2000, 2001, 1972, 2003, 2004, 1978,
       1979, 1981, 2005, 1982, 1966, 1983, 1973, 1984, 2007, 1985, 1986,
       1987, 2008, 1988, 2009, 1989, 1991, 1990, 1974, 2010, 2011, 1992,
       1993, 2002, 2012, 2013, 1994, 2014, 1975, 1977, 1976, '2004',
       '1994', '1979', '1976', '1975', '1974', '1977', '1981', '1982',
       '1983', '1984', '1978', '1980', '1968', '1976A', '1969', '1995',
       '1970', '2005', '1996', '1971', 1996, 1971], dtype=object)

There one record that include 'A' as well, so it need to be clean

In [20]:
# function
def clean(x):
    if type(x) == str:
        return x.replace('A', '')
    else:
        return x
    
df['ApprovalFY'] = df['ApprovalFY'].apply(clean).astype('int64')

### ApprovalDate & DisbursementDate

Next, I would to look at ApprovalDate and DisbursementDate. For ApprovalDate features contains information about Date SBA commitment issued while DisbursementDate features contains information about Disbursement date.

In [21]:
# Change the type of data
df['ApprovalDate'] = pd.to_datetime(df['ApprovalDate'], format='%d-%b-%y')
df['DisbursementDate'] = pd.to_datetime(df['DisbursementDate'], format='%d-%b-%y')

In [22]:
df.dtypes

LoanNr_ChkDgt                 int64
City                         object
State                        object
Zip                           int64
Bank                         object
BankState                    object
NAICS                         int64
ApprovalDate         datetime64[ns]
ApprovalFY                    int64
Term                          int64
NoEmp                         int64
NewExist                    float64
CreateJob                     int64
RetainedJob                   int64
FranchiseCode                 int64
UrbanRural                    int64
RevLineCr                    object
LowDoc                       object
DisbursementDate     datetime64[ns]
DisbursementGross            object
BalanceGross                 object
MIS_Status                   object
ChgOffPrinGr                 object
GrAppv                       object
SBA_Appv                     object
dtype: object

In [23]:
## Change some features
df = df.astype({'Zip':'str', 'NewExist':'str', 'UrbanRural':'str', 
                 'DisbursementGross':'float', 'BalanceGross':'float',
                'ChgOffPrinGr':'float', 'GrAppv':'float', 'SBA_Appv':'float'})

### NAICS

Next, I would to look at NAICS. NAICS is North American industry classification system code, the first two digits of the NAICS classification represent the economic sector

In [24]:
## Lets make sure that NAICS has 6 digits
naics_true=0
naics_false=0
for x in df['NAICS'].astype(str):
    if len(x) == 6 :
        naics_true += 1
    else:
        naics_false += 1
print(naics_true)
print(naics_false)

695480
201667


After little investigate, 201667 NAICS has 0 value so it dont has 6 digits, I will subset the data

In [25]:
df = df[df['NAICS'] != 0]

In [26]:
## first lets take first 2 digits from NAICS
df['sector'] = df['NAICS'].astype(str).apply(lambda x:x[:2])

In [27]:
## Lets encode sector feature base on the sector
df['sector'] = df['sector'].map({
    '11':'Ag_fores_fishi_hunting',
    '21':'Mining_quar_oil_gas_ext',
    '22':'Utilities',
    '23':'Construction',
    '31':'Manufacturing',
    '32':'Manufacturing',
    '33':'Manufacturing',
    '42':'Wholesale_trade',
    '44':'Retail_trade',
    '45':'Retail_trade',
    '48':'Trans_Ware',
    '49':'Trans_Ware',
    '51':'Information',
    '52':'Finance_insurance',
    '53':'RE_rental_leasing',
    '54':'Prof_scien_and_tech_serv',
    '55':'Management_and_enterprises',
    '56':'Admini_sup_and_waste_mgm_rem',
    '61':'Educational_services',
    '62':'Health_care_social_ass',
    '71':'Arts_enter_recreation',
    '72':'Accommodation_food_serv',
    '81':'Other_no_pub',
    '92':'Public_admin',
})

### FranchiseCode

Next, I would like to look at FranchiseCode feature. This feature contains information about Franchise code where (00000 or 00001) = No franchise. So, I would make a new feature to know whether the business is Franchise or No franchise 

In [28]:
## Function
def isfranchise(x):
    if x == 0 or x == 1:
        return 'No Franchise'
    else:
        return 'Franchise'
## Make a new feature
df['isFranchise'] = df['FranchiseCode'].apply(isfranchise)

### State and BankState

Next, lets look for whether or not the bank service the laon was in the same state that small business was located.

In [29]:
#Dropping Misiing value in BankState
df.dropna(subset=['BankState'], inplace=True)

In [30]:
# Create syntetic feature
df['sameState'] = np.where(df['State'] == df['BankState'], 1, 0)

### Term

Making New Features, according to the jurnal small business that has term loan more than 240 months were backed by real estate. But I would to subset the data that has Term=0, Maturities depend on the use of loan
proceeds but typically range from 5 to 25 years

In [31]:
df = df[df['Term'] !=0]

In [32]:
#New feature
df['backRealEstate'] = df['Term'].apply(lambda x:1 if x>=240 else 0)

In [33]:
df.shape

(694123, 29)

### Portion

Next, I wanna know the portion that SBA guaranteed the loan. This could be done by calculating the ratio of amount of the loan SBA guarantees and the gross amount approved by the bank

In [34]:
# New Features
df['Portion'] = df['SBA_Appv']/df['GrAppv']

The maximum SBA guarantee program is 85% for loans up to 150.000 and 75% for loans greater than 150.000. For International Trade the maximum guarantee could be 90%. Now I would checking the maximum of SBA guarantee in this the data

In [35]:
df['Portion'].describe()

count    694123.000000
mean          0.680419
std           0.175572
min           0.028140
25%           0.500000
50%           0.750000
75%           0.850000
max           1.000000
Name: Portion, dtype: float64

Since the max of guarantee for this data is 1, I would subset the data base on for Portion under 1

In [37]:
df = df[df['Portion'] <= 0.90]

In [38]:
df.shape

(647850, 30)

### NoEmp

Next, I would checking on NoEmp feature. Well, according to the SBA, a small businesses have a maximum of anywhere between 250 and 1500 employees — all depending on the specific industry the business is in

In [39]:
df['NoEmp'].describe()

count    647850.000000
mean          8.930791
std          60.588436
min           0.000000
25%           2.000000
50%           4.000000
75%           8.000000
max        9999.000000
Name: NoEmp, dtype: float64

it look like NoEmp has outliers, So I would subset the data

In [40]:
df = df[df['NoEmp'] <= 1500]

In [41]:
df.shape

(647750, 30)

### RevLineCr

In [42]:
# Revolving line of credit: Y = Yes, N = No
df['RevLineCr'].unique()

array(['N', 'Y', '0', 'T', '`', ',', nan, '1', 'C', '2', 'R', '7', '-',
       'Q'], dtype=object)

Since Revolving line of credit just contain 2 value (Yes and No), so I will subset the data

In [43]:
df = df[(df['RevLineCr'] == 'Y') | (df['RevLineCr'] == 'N')]

In [44]:
df.shape

(434164, 30)

### LowDoc

In [45]:
# LowDoc Loan Program: Y = Yes, N = No
df['LowDoc'].unique()

array(['Y', 'N', 'C', '1', nan, 'S', 'R', 'A', '0'], dtype=object)

Since LowDoc Loan Program just contain 2 value (Yes and No), so I will subset the data

In [46]:
df = df[(df['LowDoc'] == 'Y') | (df['LowDoc'] == 'N')]

In [47]:
df.shape

(429899, 30)

### NewExist

In [48]:
## NewExist
df['NewExist'].unique()

array(['2.0', '1.0', '0.0'], dtype=object)

Since NewExist just contain with 1 = Existing business, 2 = New business, I would drop others

In [49]:
df = df[(df['NewExist'] == '1.0') | (df['NewExist'] == '2.0')]

In [50]:
df.shape

(429368, 30)

### UrbanRural

In [51]:
## UrbanRural
# 1 : Urban, 2 : rural, 0 : undefined
df['UrbanRural'].unique()

array(['0', '1', '2'], dtype=object)

I would subset the data just base Urban and Rural

In [52]:
df = df[(df['UrbanRural'] == '1') | (df['UrbanRural'] == '2')]

In [53]:
df.shape

(334056, 30)

In [54]:
# checking data type after some adding syntetic feature
df.dtypes

LoanNr_ChkDgt                 int64
City                         object
State                        object
Zip                          object
Bank                         object
BankState                    object
NAICS                         int64
ApprovalDate         datetime64[ns]
ApprovalFY                    int64
Term                          int64
NoEmp                         int64
NewExist                     object
CreateJob                     int64
RetainedJob                   int64
FranchiseCode                 int64
UrbanRural                   object
RevLineCr                    object
LowDoc                       object
DisbursementDate     datetime64[ns]
DisbursementGross           float64
BalanceGross                float64
MIS_Status                   object
ChgOffPrinGr                float64
GrAppv                      float64
SBA_Appv                    float64
sector                       object
isFranchise                  object
sameState                   

In [55]:
## Change some features
df = df.astype({'sameState':'str', 'backRealEstate':'str'})

it seem like all feature has the right data type. Then we could remove some feature that would not use in the analysis.
- LoanNr_ChkDgt -- this is just Identifier Primary key
- City, State, Zip -- this is information about borrower
- Bank and BankState -- this is information about Bank that provide services loan. 
- NAICS -- replace by sector
- FranchiseCode -- replace by isFranchise
- DisbursementDate --  this feature just contain date where amount disbursed
- SBA_Appv - guaranteed amount is based on percentage of gross loan amount, replace by Portion


In [56]:
df.drop(columns=['LoanNr_ChkDgt', 'City', 'Zip', 'Bank', 'NAICS', 
                 'FranchiseCode', 'DisbursementDate', 'ApprovalDate'], inplace=True)

In [57]:
df.drop(columns=['State', 'BankState'], inplace=True)

In [58]:
df.drop(columns=['ChgOffPrinGr', 'BalanceGross'], inplace=True)

In [59]:
df.drop(columns=['SBA_Appv'], inplace=True)

In [60]:
## Checking the missing value
df.isna().sum()

ApprovalFY           0
Term                 0
NoEmp                0
NewExist             0
CreateJob            0
RetainedJob          0
UrbanRural           0
RevLineCr            0
LowDoc               0
DisbursementGross    0
MIS_Status           0
GrAppv               0
sector               0
isFranchise          0
sameState            0
backRealEstate       0
Portion              0
dtype: int64

#### Export Dataset
Save the dataframe to CSV format for the next step

In [61]:
# df.to_csv('Clean_SBA_Loan.csv', index=False)